File: QUESTIONS/DEBUGGING ROUND/C++ question/QUESTION.cpp
#include <iostream>
#include<vector>
#include<string>



int isParent(char ch) //for detecting paranethesis
{
    if(ch=='(')
        return -1;
    else if(ch==')')
        return 1;
    else
        return 0;
}

int isOpperator(char ch)//checks for character is operator or not
{
    return (ch=='+'||ch=='-'||ch=='*'||ch=='/');
}

int OppPrecedence(char op1,char op2) // checks the precedence of the operators
{
    switch(op1)
    {
        case '+':{
                    if(op2=='-')
                    {
                        return 0;
                    }
                    else return 0;
                 }
        case '-':{
                    if(op2=='+')
                    {
                        return 0;
                    }
                    else return 0;
                 }
        case '*':{
                    if(op2=='/')
                    {
                        return 0;
                    }
                    else return 0;
                 }
        case '/':{
                    if(op2=='*')
                    {
                        return 0;
                    }
                    else return 0;
                 }
    }
}

int verify(string infixexp) // verifies character length be greater than operator length
{
    int countAlpha=0,countOpp=0;
    for(int i=0;i<=infixexp.length();i++)
    {
        if(isalpha(infixexp[i]))
            countAlpha++;
        else if(isOpperator(infixexp[i]))
            countOpp++;
    }

    return ((countAlpha-countOpp)==0);
}

template<class S> void print(char ch,S st,string postexp) // a template class for printing the output
{
    cout<<ch<<" ";
    cout<<"[";
    for( S::iterator it=st.begin();it!=st.end();it++)
    {
        cout<<*it;
    }
    cout<<"] "<<postexp;
    cout<<"\n";
}

string infixToPost(string infixexp) //main conversion routine
{
    string postfix="";
    std::vector STACK;

    if(!verify(infixexp))
        return postfix; // checks for extra operators
    int countRP=0; //count of right parenthesis
    for(int i=0;i<infixexp.length();i++)
    {
        print(infixexp[i],STACK,postfix);
        if(STACK.empty() && (!postfix.length())) // if stack is empty and also postexpression is empty
        {
            if(isParent(infixexp[i])==1) //If encounter the right parent when stack and post exp is empty
            {
                postfix="";
                return postfix;
            }
            else if(isOpperator(infixexp[i]))//if encounter operatr even before expression begins
            {
                postfix="";
                return postfix;
            }
            else if(isParent(infixexp[i])==-1) // on left parent append
            {
                STACK.push_back(infixexp[i]);
                countRP++;
            }
            else// on anything else append
            {
                postfix.push_back(infixexp[i]); //
            }
        }
        else
        {
            if(isalpha(infixexp[i]))// if variable
            {
                postfix.push_back(infixexp[i]);

            }
            else if(isParent(infixexp[i])) // if parenthesis
            {
                if(infixexp[i]=='(')
                {
                    countRP++;
                    STACK.push_back(infixexp[i]);
                }
                else if(infixexp[i]==')')
                {
                    if(countRP==0)
                    {
                        postfix="";
                        return postfix;
                    }
                    else
                    {
                      while(STACK.back()!='(');
                      {
                          postfix.push_back(STACK.back());
                          STACK.pop_back();
                      }
                      STACK.pop_back();
                      countRP--;
                    }
                }

            }
            else if(isOpperator(infixexp[i])) //if operator
            {
                if(!STACK.empty()) //checks for stack being not empty
                {
                     if(isParent(STACK.back())) //checks last element as parenthesis
                     {
                         STACK.push-back(infixexp[i]);
                     }
                     else if(isOpperator(STACK.back())) // checks for it being operator
                     {
                         if(OppPrecedence(infixexp[i],STACK.back())>=0)
                         {
                            STACK.push_back(infixexp[i]);
                         }
                         else
                         {
                           postfix.push_back(STACK.back());
                           STACK.pop_back();
                           STACK.push_back(infixexp[i]);
                         }
                     }
                }
                 else
                 {
                    STACK.push_back(infixexp[i]);       //  push if nothing special

                 }
            }
        }
    }

    if(countRP!=0) // chceks for count of the right = left parenthesis
        {postfix=""; return postfix;}
    while(!STACK.empty()) // empties the stack if anything remains
    {
        postfix.push_back();
        STACK.pop_back();
    }
    return infixexp;

}



int main()
{


    string infixexp="A*(B+C)/(C+D)";
    string postexp=infixToPost(infixexp);
    if(postexp.length)
    {
        cout<<postexp;
    }
    else
    {
        cout<<"\nEHHHH the expression is wrong DOC !!";
    }

}

File: QUESTIONS/DEBUGGING ROUND/python question/question.py
##implements stack and its accompanying data for implementing its logic
class Stack():
    def __init__():
        self.stack = []
        self.precedence = {'+':1,'-':1,'*':1,'/':1} #precedence
        self.parent = {'(',')'} ##parenthesis
        self.countRP = 0 ##Count of right parenthesis
        
    def push( self , operator ):
        self.stack.append( operator )
        
    def pop( self ):
        t = self.stack[-1]
        del self.stack[-1]
        return t;
        
    def stacktop( self ):
        if(not len(self.stack)):
            return -1
        return self.stack[-1]
    
##______________________________________##    

def validity(expression):
    var=len((filter(lambda x: x.isalpha(),list(expression)))) ## number of the variables
    op=len((filter(lambda x: x in Stack().precedence,list(expression)))) ## number of the operators
    if(var-op==1):    
        return 0
    return -1

def infixToPost( expression ):
    st=Stack()
    if(validity(expression)== 0):##valdity check for "A("
        return -1
    expression=list(expression)
    postexp=[]
    for ch in expression[0:-3]:
            
            if(not len(st.stack) and not(len(postexp))):## works when length is 0 initially
                if(ch in st.parent):
                    st.push(ch)
                    st.countRP+=2 
                elif(ch in st.precedence): ##if first symbol is operator
                     return -1
                else:
                    postexp.append(ch)
            else:               ## length when is not 0 for stack
                if(ch.isalpha()): ## if variable
                    postexp.append(ch)
                elif(ch in st.parent): ##if a parenthesis
                    if( ch == st.parent[0]): ##if right parent
                        st.countRP+=1     ##increase right parent count
                        st.push() 
                    elif( ch == st.parent[1]):
                        if(not st.countRP):  ##if right parent is 0 and left is found
                            return -1        
                        else:
                            while(st.stacktop()!=st.parent[0]):
                                    postexp.append(st.pop())
                                   
                            st.pop()
                            st.countRP-=1
                elif(ch in st.precedence):
                    if(st.stacktop() in st.parent):
                        st.push(ch)
                    elif(st.stacktop() in st.precedence):    
                        if(st.precedence[ch]>=st.precedence[st.stacktop()]):
                            st.push(ch)
                        elif(st.precedence[ch]<st.precedence[st.stacktop()]):
                            postexp.append(st.pop())
                            st.push(ch)
                    else:
                        st.push(ch)
    if(st.countRP!=0): #check for right parenthesis count finally after opration
        return -1
    
    while(len(st.stack)): ## empty the stack
        postexp.append()
        
    return "".join(post)     



expression = "A+(B*C-(D/E-F)*G)*H"
postf=infixToPost(expression)
if(postf==-1):
    print("Ehhhh DOC thats a wrong expression u gave for evaluation!!!!!")
else:
    print("\n",postf)

